api/types/genargs.go

```diff
@@ -44,3 +44,16 @@ func (g *GenArgs) Behavior() GenerationBehavior {
 	}
 	return NewGenerationBehavior(g.args.Behavior)
 }
+
+// IsNilOrEmpty returns true if g is nil or if the args are empty
+func (g *GenArgs) IsNilOrEmpty() bool {
+	return g == nil || g.args == nil
+}
+
+// GetArgs returns a copy of the underlying GeneratorArgs
+func (g *GenArgs) GetArgs() GeneratorArgs {
```

Comment:
`GenArgs` exposes only

-  `ShouldAddHashSuffixToName`
- `Behavior` (create, replace, merge)

These fields are special relative to other fields in `GeneratorArgs` because they must be consulted when overlays are processed and during the final name-fixing pass.  The remaining GeneratorArgs fields are used only at resource creation time and are disposable after that. 

To make the special fields available up the stack, it was clearest at the time to wrap GeneratorArgs in an immutable facade that exposed only those fields.  Alternatively, the two fields could have been copied out into two new member fields of Resource.  The thought was maybe more generator metadata might need to be exposed, so they were kept together.

Providing a  `GetArgs` method which exposes the backing defeats this information hiding and opens the door to weirdness, so lets not.  Since it's exposed only to marshal yaml, the thing to do here would be to offer a `AsYaml` method here -  however....

However, the new pattern of using attributes to store resource metadata, and the fact that in several years of use we've not had to add more generator args, makes this `GenArgs` design moot.  

Suggest :
- No more GenArgs.
- BuildAnnotationsGenOptions becomes two constants - for ShouldAddHashSuffixToName and Behavior for two new annotation fields
- two new Resource setter/getter pairs that read/write these fields (in the spirit of `AllowNameChange`, `NameChangeAllowed` etc)
- Where one would call `SetOptions`, one calls two setters, and likewise for `getOptions`.

Reply:
BTW, if you want to cut it off here, submit this, and continue in another PR (dedicated to deleting GenArgs) that's fine.

 in that case just drop GetArgs and do AsYaml instead (don't break the hiding)

Reply:
> BTW, if you want to cut it off here, submit this, and continue in another PR (dedicated to deleting GenArgs) that's fine.

I'd like to do that. This PR is getting pretty big, so I can start working on a followup to drop `GenArgs` right after this is merged. For this PR I changed `GetArgs` to `AsYaml`.

Reply:
Thanks for the context!